### PR TITLE
Grammars scopes to config + added support to styled-components

### PR DESCRIPTION
**Pull request by mgtitimoli**
Hi there!

I've been using [Styled Components] for a while, and recently we decided to add **stylelint** to our validation pipeline.

In addition to lint from the console, I wanted to get some feedback directly in Atom, and was then when I found this package.

I installed the package, but to my surprise I was not getting any feedback, so after digging a while I notice that this was caused because the grammar for Styled Components was not included in the currently supported ones.

In order to add a new grammar there were 2 paths to take:
1. Add the new grammar directly to the [baseScopes array](https://github.com/AtomLinter/linter-stylelint/blob/master/lib/index.js#L27)
2. Promote supported grammars to be a config value, so this way users would be able to specify the grammars they want to lint

After analyzing the pros and cons I ended up picking the nice and marvelous path 2 :sparkles:.

Hope you like this :slightly_smiling_face:

### CODE REVIEW

**Comment by Arcanemagus** [comment]
Looking this over now.

File: package.json [hunk 1]
```diff
@@ -47,6 +47,33 @@
...
+        "source.css.scss",
+        "source.less",
+        "source.css.less",
+        "source.css.postcss",
+        "source.css.postcss.sugarss",
+        "source.inside-js.css.styled"
+      ],
+      "items": {
+        "type": "string",
+        "enum": [
```
Comment (Arcanemagus) [inline package.json]: Wow, although this _works_, it leads to some incredibly broken user experiences. Typing anything other than this allowed list in there results in that non-allowed item disappearing with no warning... this includes accidentally editing one of the allowed list.

There are two approaches that can be taken here:
* Have a predefined list of scopes, with individual options controlling each one, similar to how https://github.com/AtomLinter/linter-stylelint/commit/c9da67a6fa19c9961ff8ee32feb7bc938342de27 worked.
* Allow any scopes to be entered, similar to how `linter-eslint` works ([1](https://github.com/AtomLinter/linter-eslint/blob/v8.1.2/package.json#L75), [2](https://github.com/AtomLinter/linter-eslint/blob/v8.1.2/src/main.js#L29))

Since this package currently can only use the embedded `stylelint`, I would say that the first option is the better way to go for now as extra processors currently can't be added by the user (without editing the source code)

tl;dr: Change this whole option to an "Enable Styled Components Support" option 😛.

File: lib/index.js [hunk 2]
```diff
@@ -76,6 +68,9 @@ export function activate() {
   subscriptions.add(atom.config.observe('linter-stylelint.showIgnored', (value) => {
     showIgnored = value;
   }));
+  subscriptions.add(atom.config.observe('linter-stylelint.grammarScopes', (value) => {
+    grammarScopes = value;
```
Comment (Arcanemagus) [inline lib/index.js]: This setting needs to be removed, but just for your reference the implementation here wouldn't have worked anyway.

A reference to the _initial_ `grammarScopes` down in the return from `provideLinter` is the only one that is ever used, so you would have needed to update the current Array not replace it as you are doing here.

File: spec/fixtures/styled-components/bad.js [hunk 3]
```diff
@@ -0,0 +1,9 @@
+import React from 'react';
```
Comment (Arcanemagus) [inline spec/fixtures/styled-components/bad.js]: Do these `import` lines need to be in here? They are causing ESLint to blow up CI.

File: spec/linter-stylelint-spec.js [hunk 4]
```diff
@@ -301,4 +303,33 @@ describe('The stylelint provider for Linter', () => {
       );
     });
   });
+
+  describe('works with JS files (language-babel) with Styled Components and', () => {
+    beforeEach(() => {
+      atom.config.set('linter-stylelint.disableWhenNoConfig', false);
```
Comment (Arcanemagus) [inline spec/linter-stylelint-spec.js]: There _is_ a configuration for those files, is there any reason you are disabling this setting?

File: spec/linter-stylelint-spec.js [hunk 5]
```diff
@@ -301,4 +303,33 @@ describe('The stylelint provider for Linter', () => {
...
+      waitsForPromise(() => atom.packages.activatePackage('language-babel'));
```
Comment (Arcanemagus) [inline spec/linter-stylelint-spec.js]: Does this _need_ `language-babel`? I'd like to avoid bringing that in here if possible. If it is _required_, then it needs to be added to the list of pre-installed packages in the CI scripts.

File: spec/linter-stylelint-spec.js [hunk 6]
```diff
@@ -301,4 +303,33 @@ describe('The stylelint provider for Linter', () => {
...
+    });
+
+    it('correctly detects an error present in a file', () => {
+      const ntscMessage = 'Expected a trailing semicolon (<a href="http://stylelint.io/user-guide/rules/declaration-block-trailing-semicolon">declaration-block-trailing-semicolon</a>)';
```
Comment (Arcanemagus) [inline spec/linter-stylelint-spec.js]: I'm not sure why ESLint isn't blowing up on this line being too long already, but can you split this into two lines?

**Comment by Arcanemagus** [comment]
@mgtitimoli Are you interested in continuing work on this?

**Comment by mgtitimoli** [comment]
Hi @Arcanemagus,

Sorry for taking too long to answer, I will get into this during the day.

**Comment by bcotrim** [comment]
I would find this useful for my projects.
@mgtitimoli are you going to finish this up? otherwise I could pick this up.

**Comment by zaaack** [comment]
Any updates?

**Comment by konstantin24121** [comment]
Anybody complete this PR. Sadly lose linting in Atom with styled-components

**Comment by Arcanemagus** [comment]
This PR is now rather out of date, I'll be re-implementing this after the other open PR's are finished up if nobody beats me to it 😉.

**Comment by Ky6uk** [comment]
I found there is no way to lint styled components through this linter yet. Isn't there?
